test/utils_test.js

```diff
@@ -603,4 +627,23 @@ describe("utils", () => {
       expect(pad(4, 3)).eql("004");
     });
   });
+
+  describe("toErrorList()", () => {
+    it("should convert an errorSchema into a flat list", () => {
+      expect(toErrorList({
+        a: {
+          b: {
+            __errors: ["err1", "err2"]
+          }
+        },
+        c: {
+          __errors: ["err3"]
+        }
+      })).eql([
+        {stack: "b err1"},
+        {stack: "b err2"},
```

Comment:
> Well, this output format no longer corresponds to the `jsonschema` return type.

I don't want to stick to the format exposed by *jsonschema* too much, mostly because I'm not finding it super convenient as it uses custom error class instances, and because I'd like to decouple our implementation from their API as much as I can. For example, *jsonschema* doesn't allow defining our very own error messages instead of their builtins (eg. see #155), and we may want to investigate other libraries at some point.

> Specifically, in `jsonschema` it would be `instance.a.b`, and there would be properties called `property`, `name`, `instance`, etc. 

... and `stack`, indeed, which is the *jsonschema* property name they use for storing the complete error message. Hence why I'm complying here.

> I think we should either match that output or document the format we're producing instead. 

We must define our API and associated format before documenting it; be sure I'll update docs as soon as the ongoing implementation is validated :)

> I don't have a very clear sense of what the use case is for the error list, so I don't know which is better. 

The use case is for displaying the list of all errors at the top of the form, in a flat fashion:

![](http://i.imgur.com/omaKEw4.png)

This capture is made against [the hosted demo](http://mozilla-services.github.io/react-jsonschema-form/) (see the *Errors* tab).

As you can see, iterating over the error list is pretty much straigthforward to render this. Basically we really do have the two distinct use cases of

1. rendering a flat list of errors at the top of the form (our `errors` list);
2. rendering errors contextually to their field level, which needs a tree (our `errorSchema` map).

> It looks like you intend for `stack` to be usable directly as an error message, so maybe we should call it `errorMsg`.

No because as mentioned previously, I need to comply with the *jsonschema* format which uses a `stack` property to store the full error description string.

> I think we will probably find that having the path available is going to be useful eventually even if it doesn't show up in the error message (that's why `jsonschema` provides it).

Yeah that's why I still need to think about it. My main concern is we're already processing a lot of this stuff from one format to another and vice versa. We could invent some intermediary format, which is likely to look a lot like the `{errors, errorSchema}` one we already have (as, again, we actually need both data structures). Just need to ensure we're not wasting too many CPU cycles going back and forth here.